submissions/antonkottans/tjsw-post-oop/index.js

```diff
@@ -0,0 +1,185 @@
+const KEYS_VALUE_SEPARATOR = ": ";
+const PROPS_SEPARATOR = "; ";
+
+class Creature {
+    constructor({ name, gender, friends = [], species, saying }) {
+        this.name = name;
+        this.species = species;
+        this.gender = gender;
+        this.friends = friends;
+        this.saying = saying;
+    }
+
+    createArrayOfProps() {
+        return ["name", "species", "gender", "saying", "friends"]
+            .map((key) => {
+                if (key != "friends")
+                    return key + KEYS_VALUE_SEPARATOR + this[key];
+                return this.friends.length
+                    ? "friends: " +
+                          this.friends.map((friend) => friend.name).join(", ")
+                    : "";
+            })
+            .filter((_) => _);
+    }
+
+    addFriends(...newFriends) {
+        if (newFriends.length > 0)
+            this.friends = [...this.friends, ...newFriends];
+    }
+}
+
+class Animal extends Creature {
+    constructor({ name, gender, friends, paws, species, saying }) {
+        super({ name, gender, friends, species, saying });
+        this.paws = paws;
+    }
+
+    createArrayOfProps() {
+        return [
+            ...super.createArrayOfProps(),
+            "paws" + KEYS_VALUE_SEPARATOR + this["paws"],
+        ];
+    }
+}
+
+class Human extends Creature {
+    constructor({ name, gender, friends, legs, hands, saying }) {
+        super({ name, gender, friends, species: "human", saying });
+        this.legs = legs;
+        this.hands = hands;
+    }
+
+    createArrayOfProps() {
+        return [
+            ...super.createArrayOfProps(),
+            ...["legs", "hands"].map(
+                (key) => key + KEYS_VALUE_SEPARATOR + this[key]
+            ),
+        ];
+    }
+}
+
+class Dog extends Animal {
+    constructor({
+        name,
+        gender,
+        friends,
+        paws,
+        tail = "ordinary tail",
+        saying,
+    }) {
+        super({ name, gender, friends, paws, tail, species: "dog", saying });
+        this.tail = tail;
```

Comment:
Looks like Dog and Cat have tails. Why this property is not generalized in some base class?

Reply:
becouse not all animals has tail, like frogs

Reply:
Do we have any tailless animals yet?
We have suboptimal code now already in favour of possible need that we do not have yet.
Pre-mature optimization is a poor practice.
Most likely if we have tailless animals we would still need split Animal class for tailed and tailless. 
So, we need a class that holds a shared property anywa.